presto-accumulo/src/main/java/com/facebook/presto/accumulo/AccumuloClient.java

```diff
@@ -0,0 +1,1177 @@
+                    // Get the bytes of the key
+                    byte[] keyBytes = kvp.getKey().getRow().copyBytes();
+
+                    // If the last byte is <, then we have hit the default tablet, so use this location
+                    if (keyBytes[keyBytes.length - 1] == '<') {
+                        location = Optional.of(kvp.getValue().toString());
+                        break;
+                    }
+                    else {
+                        // Chop off some magic nonsense
+                        scannedCompareKey.set(keyBytes, 3, keyBytes.length - 3);
+
+                        // Compare the keys, moving along the tablets until the location is found
+                        if (scannedCompareKey.getLength() > 0) {
+                            int compareTo = splitCompareKey.compareTo(scannedCompareKey);
+                            if (compareTo <= 0) {
+                                location = Optional.of(kvp.getValue().toString());
+                            }
+                            else {
+                                // all future tablets will be greater than this key
+                                break;
+                            }
+                        }
+                    }
+                }
+                scan.close();
+            }
+
+            // If we were unable to find the location for some reason, return the default tablet
+            // location
+            return location.isPresent() ? location : getDefaultTabletLocation(table);
+        }
+        catch (Exception e) {
+            // Swallow this exception so the query does not fail due to being unable
+            // to locate the tablet server for the provided Key.
+            // This is purely an optimization, but we will want to log the error.
+            LOG.error("Failed to get tablet location, returning dummy location", e);
+            e.printStackTrace();
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Gets the location of the default tablet for the given table
+     *
+     * @param fulltable Fully-qualified table name
+     * @return TabletServer location of the default tablet
+     */
+    private Optional<String> getDefaultTabletLocation(String fulltable)
+    {
+        try {
+            // Get the table ID
+            String tableId = conn.tableOperations().tableIdMap().get(fulltable);
+
+            // Create a scanner over the metadata table, fetching the 'loc' column of the default
+            // tablet row
+            Scanner scan = conn.createScanner("accumulo.metadata",
+                    conn.securityOperations().getUserAuthorizations(conf.getUsername()));
+            scan.fetchColumnFamily(new Text("loc"));
+            scan.setRange(new Range(tableId + '<'));
+
+            // scan the entry
+            Optional<String> location = Optional.empty();
+            for (Entry<Key, Value> kvp : scan) {
+                if (location != null) {
```

Comment:
This does not look good.
`Optional.empty()` is also non-null.
Maybe you want to check if `location.isPresent()`?